### ISSUE TITLE
Possible mistake in GWAS demo notebooks

### ISSUE THREAD

**Issue by eric-czech** [projectglow/glow#257]
I think this step in the [WGR demo notebook](https://glow.readthedocs.io/en/latest/_static/notebooks/tertiary/glowgr.html) has a mistake:

```python
wgr_gwas = variant_df.join(adjusted_phenotypes, ['contigName']).select(
  ...
  expand_struct(linear_regression_gwas( 
    variant_df.values,
    adjusted_phenotypes.values,
    # ** This appears to get flattened as row-major and then reshaped as column-major internally **
    lit(covariates.to_numpy())
  )))
 ```

Here's a short program to show what made me think that:

```python
import glow
from glow import linear_regression_gwas, expand_struct
import numpy as np
from pyspark.ml.linalg import DenseMatrix
from pyspark.sql.session import SparkSession
from pyspark.sql import Row
from pyspark.sql import functions as F

spark = SparkSession.builder\
    .config('spark.jars.packages', 'io.projectglow:glow_2.11:0.5.0')\
    .config("spark.sql.execution.arrow.pyspark.enabled", "true")\
    .getOrCreate()
glow.register(spark)

# Create arrays to test with
np.random.seed(0)
x = np.array([ # Covariates
    [1, -1],
    [2, -2],
    [3, -3],
    [4, -4.],
])
g = np.array([0., 1., 2., 0.])
b = np.array([0., 1.])
# The `beta` for g should be very close to 1
y = g + np.dot(x, b) + np.random.normal(scale=.01, size=g.size)

#### Version 1
# Pass the covariates directly with correct ordering
dm = DenseMatrix(numRows=x.shape[0], numCols=x.shape[1], values=x.ravel(order='F').tolist())
np.testing.assert_equal(x, dm.toArray()) # `dm` is definitely defined correctly
spark.createDataFrame([Row(genotypes=g.tolist(), phenotypes=y.tolist(), covariates=dm)])\
    .select(expand_struct(linear_regression_gwas('genotypes', 'phenotypes', 'covariates')))\
    .show()
+------------------+--------------------+--------------------+
|              beta|       standardError|              pValue|
+------------------+--------------------+--------------------+
|0.9948866298677059|0.001300287478677...|8.320427901051293E-4|
+------------------+--------------------+--------------------+
# *Correct result*

#### Version 2
# Pass the covariates raveled as row-major and let DenseMatrix (incorrectly)
# interpret the concatenated rows as columns
dm = DenseMatrix(numRows=x.shape[0], numCols=x.shape[1], values=x.ravel(order='C').tolist())
print(dm.toArray())
[[ 1.  3.]
 [-1. -3.]
 [ 2.  4.]
 [-2. -4.]]
spark.createDataFrame([Row(genotypes=g.tolist(), phenotypes=y.tolist(), covariates=dm)])\
    .select(expand_struct(linear_regression_gwas('genotypes', 'phenotypes', 'covariates')))\
    .show()
+------------------+-------------------+--------------------+
|              beta|      standardError|              pValue|
+------------------+-------------------+--------------------+
|-2.382793056158103|0.20221757585071454|0.053898082087957454|
+------------------+-------------------+--------------------+
# * Incorrect result, beta not ~= 1

#### Version 3
# Pass the covariates array as a literal the same way it is done in the demo notebook
spark.createDataFrame([Row(genotypes=g.tolist(), phenotypes=y.tolist())])\
    .select(expand_struct(linear_regression_gwas('genotypes', 'phenotypes', F.lit(x))))\
    .show()
+------------------+-------------------+--------------------+
|              beta|      standardError|              pValue|
+------------------+-------------------+--------------------+
|-2.382793056158103|0.20221757585071454|0.053898082087957454|
+------------------+-------------------+--------------------+
# * Incorrect result when doing what the notebook does, beta not ~= 1
```

Version 3 using `F.lit(x)`, like the demo notebook, is seeing the same incorrectly ordered covariates as version 2 (and produces the wrong result).  Perhaps there is a better way to fix it, but `F.lit(covariates.to_numpy())` should actually be something like`F.lit(covariates.to_numpy().T.ravel(order='C').reshape(covariates.shape))`.  Here's an example that works as expected:

```python
x_weird = x.T.ravel(order='C').reshape(x.shape)
print(x_weird)
spark.createDataFrame([Row(genotypes=g.tolist(), phenotypes=y.tolist())])\
    .select(expand_struct(linear_regression_gwas('genotypes', 'phenotypes', F.lit(x_weird))))\
    .show()
[[ 1.  2.]
 [ 3.  4.]
 [-1. -2.]
 [-3. -4.]]
+------------------+--------------------+--------------------+
|              beta|       standardError|              pValue|
+------------------+--------------------+--------------------+
|0.9948866298677059|0.001300287478677...|8.320427901051293E-4|
+------------------+--------------------+--------------------+
```

Hopefully there's a more graceful way to handle that internally.

**Comment by eric-czech** [comment]
It looks like this affects the linear and logistic regression examples too.

**Comment by eric-czech** [comment]
Ah, I think the issue is here in `TwoDimensionalDoubleNumpyArrayConverter ` ([conversions.py](https://github.com/projectglow/glow/blob/master/python/glow/conversions.py#L92)):

```python
flat_arr = object.ravel() # This flattens columns first but DenseMatrix expects the other order
java_arr = _convert_numpy_to_java_array(flat_arr)
dense_matrix = sc._jvm.org.apache.spark.ml.linalg.DenseMatrix(
  object.shape[0], object.shape[1], java_arr)
``` 

A more concise version of the issue:

```python
x = np.column_stack([np.arange(4), -np.arange(4, dtype=float)])
r = spark.createDataFrame(pd.DataFrame({'i': [0]})).withColumn('x', F.lit(x)).limit(1).collect()[0]
print(x)
[[ 0. -0.]
 [ 1. -1.]
 [ 2. -2.]
 [ 3. -3.]]
print(r)
Row(i=0, x=DenseMatrix(4, 2, [0.0, -0.0, 1.0, -1.0, 2.0, -2.0, 3.0, -3.0], False))                                                                                                                                                                      
print(r.x.toArray())
[[ 0.  2.]
 [-0. -2.]
 [ 1.  3.]
 [-1. -3.]]
```

**Comment by kianfar77** [comment]
Thanks a lot @eric-czech for raising this. Fixed in #258
